ament_copyright/ament_copyright/__init__.py

```diff
@@ -12,31 +12,354 @@
+        print('No errors')
+        rc = 0
+    else:
+        print('%d errors' % error_count, file=sys.stderr)
+        rc = 1
+
+    # generate xunit file
+    if args.xunit_file:
+        testname = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if testname.endswith(suffix):
+            testname = testname[0:-len(suffix)]
+        xml = get_xunit_content(report, testname, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+
+    return rc
+
+
+def get_copyright_names():
+    names = {}
+    for entry_point in pkg_resources.iter_entry_points(
+            group=AMENT_COPYRIGHT_GROUP):
+        name = entry_point.load()
+        names[entry_point.name] = name
+    return names
+
+
+def get_files(paths, extensions):
+    files = []
+    for path in paths:
+        if os.path.isdir(path):
+            for dirpath, dirnames, filenames in os.walk(path):
+                # ignore folder starting with . or _
+                dirnames[:] = [d for d in dirnames if d[0] not in ['.', '_']]
+                dirnames.sort()
+
+                # select files by extension
+                for filename in sorted(filenames):
+                    _, ext = os.path.splitext(filename)
+                    if ext in ['.%s' % e for e in extensions]:
+                        files.append(os.path.join(dirpath, filename))
+        if os.path.isfile(path):
+            files.append(path)
+    return files
+
+
+def add_missing(filenames, name):
+    copyright = 'Copyright %s %s' % (time.strftime('%Y'), name)
+    print('Adding the following copyright notice:')
+    print(' ', copyright)
+    print('To the following file:')
+
+    for filename in filenames:
+        has_copyright = check_file_for_copyright(filename)
+        if has_copyright:
+            continue
+
+        print('*', filename)
+        add_copyright(filename, copyright)
+
+
+def add_copyright_year(filenames):
+    current_year = time.strftime('%Y')
+    print('Adding the current year to existing copyright notices:')
+
+    # regex for matching years or year ranges (yyyy-yyyy) separated by colons
+    year = '\d{4}'
+    year_range = '%s-%s' % (year, year)
+    year_or_year_range = '(%s|%s)' % (year, year_range)
+    pattern = '(\s+%s(\s*,\s*%s)*)\s+.+' % \
+        (year_or_year_range, year_or_year_range)
+    regex = re.compile(pattern)
+
+    for filename in filenames:
+        has_copyright = check_file_for_copyright(filename)
+        if not has_copyright:
+            continue
+
+        with open(filename, 'r') as h:
+            content = h.read()
+
+        index = scan_past_coding_and_shebang_lines(content)
+        index = scan_past_empty_lines(content, index)
+
+        assert is_copyright_line(content, index), \
+            "Could not find copyright line in file '%s'" % filename
```

Comment:
This is a good question we should probably answer in a generic way. I usually use `assert` when it is unexpected to happen. Showing the stacktrace to the use should be fine in that case since it should "never" be the case. But thats just my 2 cent.